File: solutions/kotlin/circular-buffer/1/src/main/kotlin/CircularBuffer.kt
import kotlin.collections.ArrayDeque

// TODO: implement proper exceptions to complete the task
class EmptyBufferException : Exception("Empty buffer") 
class BufferFullException : Exception("Buffer full")

class CircularBuffer<T>(val _size: Int) {
    var size: Int = _size
    var read_index: Int = 0;
    var write_index: Int = 0
    // type here is Any__?__ to allow for Null valued array
    var buffer: Array<Any?> = arrayOfNulls(_size)

    fun is_empty(): Boolean {
        return (this.write_index <= this.read_index);
    }

    fun is_full(): Boolean {
        return (this.write_index - this.read_index) >= this.size;
    }
    
    fun read() : T {
        if (this.is_empty()) {
            throw EmptyBufferException();
        }
        var value: T = buffer[this.read_index % this.size] as T;
        this.read_index += 1;
        return value;
    }

    fun write(value: T) {
        if (this.is_full()) {
            throw BufferFullException();
        }
        this.buffer[this.write_index % this.size] = value as Any?;
        this.write_index += 1;

    }

    fun overwrite(value: T) {
        if (! this.is_full()) {
            // no need to overwrite as the thing still has some space
            this.write(value);
            return;
        } 

        this.buffer[this.read_index % this.size] = value as Any?;
        this.read_index += 1;
        this.write_index += 1;
        
    }

    fun clear() {
        if (this.is_empty()) {
            return
        }

        this.read_index = this.write_index;
    }
}

File: solutions/csharp/circular-buffer/1/CircularBuffer.cs
public class CircularBuffer<T>
{
    private T[] buffer;
    private int write_index = 0;
    private int read_index = 0;
    private bool cleared = true;
    public CircularBuffer(int capacity)
    {
        
        buffer = new T[capacity];
    }

    public T Read()
    {
        if (cleared) {
            throw new InvalidOperationException("empty buffer can't be accessed");
        }
            
        int _read_index = read_index % buffer.Length;
        read_index += 1;

        // we have read all entries
        if (read_index == write_index) {
            cleared = true;
        }

        return buffer[_read_index];
    }

    public void Write(T value)
    {
        // check if full
        if ((write_index - read_index) >= buffer.Length) {
            throw new InvalidOperationException("Full buffer!");
        }
        
        buffer[write_index % buffer.Length] = value;
        write_index = write_index + 1;
        cleared = false;
    }

    public void Overwrite(T value)
    {
        // check if not full
        if ((write_index - read_index) < buffer.Length) {
            Write(value);
            return;
        }
        
        buffer[read_index % buffer.Length] = value;
        read_index = read_index + 1;
        write_index = write_index + 1;
    }

    public void Clear()
    {
        write_index = 0;
        read_index = 0;
        cleared = true;
    }
}